### PR TITLE
Use gas price from pending transaction if detected

### PR DESCRIPTION
**Pull request by vkgnosis**
This is a partial fix for https://github.com/gnosis/gp-v2-services/issues/445 . It should work fine in case of restarts but it might still print an error if the pending transactions gets mined first. It can also help against the case where the node forgets a transaction.

A more proper solution should go into the transaction-retry crate but that's harder to figure out how to do so I'd like to have this in the mean time.

### Test Plan
Hard to test until we see this actually happening. I did test the getBlock(Pending) call to see that it really returns pending transactions and I saw that the total gas use of pending transactions was well above 12 mil so it seems to be bigger than just "one" pending block.

### CODE REVIEW

File: solver/src/settlement_submission.rs [hunk 1]
```diff
@@ -143,3 +172,19 @@ async fn tenderly_link(
...
+) -> Result<Option<U256>> {
+    let block = web3
+        .eth()
+        .block_with_txs(BlockId::Number(BlockNumber::Pending))
+        .await?
+        .ok_or_else(|| anyhow!("empty block"))?;
+    let transaction = block
+        .transactions
+        .iter()
+        .find(|transaction| transaction.from == *address);
```
Comment (fleupold) [inline solver/src/settlement_submission.rs:188]: Should we also check nonce here to make sure we are finding the right transaction?

File: solver/src/settlement_submission.rs [hunk 2]
```diff
@@ -31,13 +33,22 @@ pub async fn submit(
...
 
+    let nonce = transaction_count(contract)
+        .await
+        .context("failed to get transaction_count")?;
+    let address = match &contract.defaults().from {
+        Some(account) => account.address(),
+        None => return Err(anyhow!("no default sender address")),
+    };
+    let web3 = contract.raw_instance().web3();
+    let pending_gas_price = recover_gas_price_from_pending_transaction(&web3, &address)
```
Comment (fleupold) [inline solver/src/settlement_submission.rs:48]: nit: Do we need to fail in case we cannot recover the pending transaction? Would it not be "safe" to assume there is none?

File: solver/src/settlement_submission.rs [hunk 3]
```diff
@@ -31,13 +33,22 @@ pub async fn submit(
...
     let settlement = encode_settlement(&settlement)?;
     // Check that a simulation of the transaction works before submitting it.
     simulate_settlement(&settlement, contract).await?;
 
+    let nonce = transaction_count(contract)
+        .await
+        .context("failed to get transaction_count")?;
+    let address = match &contract.defaults().from {
+        Some(account) => account.address(),
+        None => return Err(anyhow!("no default sender address")),
```
Comment (fleupold) [inline solver/src/settlement_submission.rs:45]: Curious on why error here instead of crash? Wouldn't it be a non-recoverable configuration error if from is not set?

File: solver/src/settlement_submission.rs [hunk 4]
```diff
@@ -55,7 +66,25 @@ pub async fn submit(
...
+    // transaction gets mined first in which case our new transaction would fail with "nonce already
+    // used".
+    if let Some(gas_price) = pending_gas_price {
+        tracing::info!(
+            "detected existing pending transaction with gas price {}",
+            gas_price
+        );
+        let gas_price = gas_price.to_f64_lossy();
+        let gas_price = transaction_retry::gas_price_increase::minimum_increase(gas_price);
+        stream = futures::stream::once(std::future::ready(gas_price))
```
Comment (fleupold) [inline solver/src/settlement_submission.rs:84]: Isn't retry expecting a monotonically increasing gas price stream? Here is looks as if it was possible to have the previous (potentially high value, followed by a lower value).

Maybe this would be a good moment to refactor `gas_price_stream` to a proper struct which can be composed like:

```rust
let stream = GasPriceStream::new(target_confirm_time, gas_limit, estimator)
  .capped(gas_price_cap);

if let Some(gas_price) = pending_gas_price {
  stream = stream.previous_value(gas_price)
}

stream = stream.minimum_increase();
```

Of course this can be done as functions as well.

**Comment by fleupold** [comment]
As for testing, I think it's reasonable for this case to do a local mainnet test (with the testing ETH we have) and stop and restart the solver within a short amount of time. Could hardcode the gas price to be somewhat low to be sure that it isn't mined too quickly.

File: solver/src/settlement_submission.rs (hunk 3 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs:45]: I wasn't sure what was more appropriate. Changed it to panic now.

File: solver/src/settlement_submission.rs (hunk 2 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs:48]: An error in that function means communication with the node failed which is very different from no gas price so I prefer it working this way.

File: solver/src/settlement_submission.rs (hunk 1 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs:188]: The nonce would only be wrong if there are multiple pending transactions right? I felt we could ignore that case for now because it should never happen. If we wanted to handle it we could make sure to always take over the lowest pending nonce.

File: solver/src/settlement_submission.rs (hunk 4 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs]: Oh right, for some reason I thought the enforce_min_increase adapter would handle this but obviously it is only applied to the second stream. I have changed `gas_price_stream` to take another parameter which passes an optional intial gas price. Will think about refactoring it more now.

**Comment by vkgnosis** [comment]
That's a good idea for a test. Will do it once PR is approved.

File: solver/src/settlement_submission.rs (hunk 1 above)
Comment (fleupold) [reply solver/src/settlement_submission.rs:193]: We need to outbid the gas price of the transaction with the same nonce so even if we had more than one pending tx I think the nonce == check would be correct (instead of checking the gas price of the lowest pending nonce):

> "Transaction gas price supplied is too low. There is another transaction with same nonce in the queue. Try increasing the gas price or incrementing the nonce.",

It's true that currently this is not necessary but it feels like a very small addition to pass the nonce and do the check. Given this code path is already an edge case it is risky to forget/not test this if we want to handle multiple parallel settlement transactions.

File: solver/src/settlement_submission.rs [hunk 5]
```diff
@@ -31,13 +33,22 @@ pub async fn submit(
...
-        .await
-        .context("failed to get transaction_count")?;
     let settlement = encode_settlement(&settlement)?;
     // Check that a simulation of the transaction works before submitting it.
     simulate_settlement(&settlement, contract).await?;
 
+    let nonce = transaction_count(contract)
+        .await
+        .context("failed to get transaction_count")?;
+    let address = match &contract.defaults().from {
```
Comment (fleupold) [inline solver/src/settlement_submission.rs:43]: nanonit: why not expect?

File: solver/src/settlement_submission.rs (hunk 2 above)
Comment (fleupold) [reply solver/src/settlement_submission.rs:48]: I think it's good that the method returns Result<Option> just not sure an error should shortcut and fail the runloop. We could log the error and continue assuming there is no pending transaction. But we can also do that if we see this happening.

**Review by fleupold — APPROVED** [review]
Looks good!

**Review by nlordell — APPROVED** [review]

File: solver/src/settlement_submission.rs (hunk 5 above)
Comment (nlordell) [reply solver/src/settlement_submission.rs:43]: I would also slightly prefer expect here.

File: solver/src/settlement_submission.rs (hunk 2 above)
Comment (fleupold) [reply solver/src/settlement_submission.rs:48]: <a href="https://gitme.me/image?url=https%3A%2F%2Fmedia2.giphy.com%2Fmedia%2FQW4aFaN0GSLshdzEtA%2Fgiphy-downsized-medium.gif&token=metoo" data-gitmeme-token="metoo"></a>

File: solver/src/settlement_submission.rs (hunk 1 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs:193]: added

File: solver/src/settlement_submission.rs (hunk 5 above)
Comment (vkgnosis) [reply solver/src/settlement_submission.rs:43]: The reason was that this is a reference but unwrap/exect consume so can't use them. Changed it to clone + expect.

**Comment by vkgnosis** [comment]
Still have to test this with mainnet before merging.

**Comment by vkgnosis** [comment]
Tested on mainnet by creating a low gas price transaction, restarting the binary and seeing that it gets discovered. Then I printed the gas price out to see that it first uses previous_gas_price * 12.5 and then goes back to the estimator. For the test I just did noop transactions to save on gas.
